lib/volk_prefs.c

```diff
@@ -63,6 +64,48 @@ void volk_get_config_path(char* path, bool read)
     return;
 }
 
+
+static struct volk_preferences {
+    volk_arch_pref_t* volk_arch_prefs;
+    size_t n_arch_prefs;
+    atomic_int initialized;
+
+} volk_preferences;
+
+
+void volk_initialize_preferences()
+{
+    if (!atomic_fetch_and(&volk_preferences.initialized, 1)) {
```

Comment:
To be honest, I am not very well versed the usage of atomic variables, but I do believe that this code is flawed:
* Being a `static` variable with no initializer provided, `volk_preferences` will be zero initialized. This means that `volk_preferences.initialized` will stay `0` for ever, i.e. the preferences will be loaded upon every call.
* Other threads must be prevented from loading prefences while some thread loads them and they must be notified when the load is completed. Two states are not sufficient to implement this, three are required (as far as I can tell):
  * state 0: Not initialized so far. Any thread observing this state changes state to 1. Note that the state switch – load of state, comparison to 0 and a possible modification of state – must be performed atomically => atomic compare exchange.
  * state 1: Some thread is loading the configuration. Any other threads must wait until this operation is finished. The loading thread signals this by setting state to 2.
  * state 2: Configration is loaded.

Reply:
`volk_preferences.initialized` should be set to `1` with `atomic_fetch_and(&volk_preferences.initialized, 1)`. It's not a constant after all.

I totally agree with your stages. I'd like to add a mutex here but unlike `volk_preferences.initialized`, default initialization is insufficient here. And I don't know how to initialize a mutex without user interaction.